File: ProiectSemestru/src/main/java/project/moto/Repository/DatasbaseRepository/TeamDBRepository.java
package project.moto.Repository.DatasbaseRepository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import project.moto.Domain.Team;
import project.moto.Repository.TeamRepository;
import project.moto.Utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.util.*;

public class TeamDBRepository implements TeamRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    public TeamDBRepository(Properties props) {
        this.dbUtils = new JdbcUtils(props);
        logger.info("Initialized TeamDBRepository with properties: {}", props);

    }

    /**
     * Find the Team with the given id
     * @param tID -the id of the team to be returned
     * id must not be null
     * @return Team - null if it doesnt exist
     */
    @Override
    public Optional<Team> findOne(Integer tID) {
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("select * from Team where id = ?")) {
            preStmt.setInt(1, tID);
            try (var result = preStmt.executeQuery()) {
                if (result.next()) {
                    int id = result.getInt("Id");
                    String name = result.getString("Name");
                    Team team = new Team(name);
                    team.setId(id);
                    logger.traceExit("Team Found");
                    return Optional.of(team);
                }
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
        logger.traceExit("Team not found");
        return Optional.empty();
    }

    /**
     * Find all Teams
     * @return Map<Integer,Team> - all Teams
     */
    @Override
    public Map<Integer,Team> findAll() {
        logger.info("Finding Teams from database");
        Connection con = dbUtils.getConnection();
        Map<Integer,Team> data = new HashMap<>();
        try (var preStmt = con.prepareStatement("select * from Team")) {
            try (var result = preStmt.executeQuery()) {
                while (result.next()) {
                    int id = result.getInt("id");
                    String name = result.getString("Name");
                    Team team = new Team(name);
                    team.setId(id);
                    data.put(id,team);
                }
                logger.traceExit("Found {} Teams",data.size());
                return data;
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Save a Team to the database
     * @param entity
     * entity must be not null
     * @return Optional<Team> - null if the Team was not successfully saved
     */
    @Override
    public Optional<Team> save(Team entity) {
        logger.traceEntry("saving team {} ", entity.getId());
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement(
                "insert into Team (Name) values (?)",PreparedStatement.RETURN_GENERATED_KEYS)) {
            preStmt.setString(1, entity.getName());
            var result = preStmt.executeUpdate();
            if (result > 0) {
                try (var generatedKeys = preStmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entity.setId(generatedKeys.getInt(1));
                    }
                }
                logger.traceExit("Team {} saved successfully", entity.getId());
                return Optional.of(entity);
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
        logger.traceExit("Team {} not saved", entity.getId());
        return Optional.empty();
    }

    /**
     * Delete a Team from the database
     * @param id
     * id must be not null
     * @return Optional<Team> - null if the Team was not successfully deleted
     */
    @Override
    public Optional<Team> delete(Integer id) {
        logger.traceEntry("deleting team {} ", id);
        Connection con = dbUtils.getConnection();
        Optional<Team> team = findOne(id);
        if(team.isPresent()) {
            try (var preStmt = con.prepareStatement("delete from Team where id = ?")) {
                preStmt.setInt(1, id);
                preStmt.executeUpdate();
                logger.traceExit("Deleted team {}", id);
                return team;
            } catch (Exception e) {
                logger.error(e.getMessage());
                throw new RuntimeException(e);
            }
        }else{
            logger.traceExit("Team not found");
            return Optional.empty();
        }
    }

    /**
     * Update a Team in the database and locally
     * @param entity - entity must not be null
     * @return Optional<Team> - null if the Team was not successfully updated
     */
    @Override
    public Optional<Team> update(Team entity) {
        logger.traceEntry("Updating team {} ", entity.getId());
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("update Team set Name = ? where id = ?")) {
            preStmt.setString(1, entity.getName());
            preStmt.setInt(2, entity.getId());
            preStmt.executeUpdate();
            logger.traceExit("Successfully updated team {} data", entity.getId());
            return Optional.of(entity);

        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
    }
}

File: ProiectMpp/ProiectMpp/Repository/DatabaseRepository/TeamDbRepository.cs
using System.Data;
using log4net;
using ProiectMpp.Domain;
using log4net.Util;
using ProiectMpp.ConnectionUtills;

namespace ProiectMpp.Repository.DatabaseRepository;

public class TeamDbRepository: ITeamRepository
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(TeamDbRepository));
    private readonly IDbConnection _connection;
    public TeamDbRepository(string connectionString)
    {
        _connection = ConnectionFactory.getInstance().createConnection(connectionString);
        Log.Info("Initializing PlayerDBRepository ");
    }

     
    public Team? FindOne(int id)
    {
        Log.Info($"Finding team with id {id}");

        try
        {
            _connection.Open();
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Team WHERE id = @id";
            var parameter = cmd.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            cmd.Parameters.Add(parameter);
            using IDataReader reader = cmd.ExecuteReader();
            if (reader.Read())
            {
                Team t = new Team(reader.GetString(0));
                t.Id = reader.GetInt32(1);
                Log.InfoExt("Team found: " + t);
                return t;
            }
            return null;
        }
        catch (Exception e)
        {
                Log.Error(e);
                throw new Exception(e.Message);
        }
        finally
        {
                _connection.Close();
        }
        
    }
     
    public IDictionary<int,Team> FindAll()
    {
        Log.Info("Retrieving all teams");
        Dictionary<int,Team> data = new Dictionary<int,Team>();
        try
        {
            _connection.Open();
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Team";
            using IDataReader reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                Team t = new Team(reader.GetString(0));
                t.Id = reader.GetInt32(1);
                data.Add(t.Id, t);
            }
            Log.InfoExt("Teams found: " + data.Count.ToString());
            return data;

        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            _connection.Close();
        }
        
    }
    
    
    public Team Save(Team entity)
    {
        Log.Info($"Saving team with id {entity.Id}");
        try
        {
            _connection.Open();
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "INSERT INTO Team (Name) VALUES (@name)";
            var parameter = cmd.CreateParameter();
            parameter.ParameterName = "@name";
            parameter.Value = entity.Name;
            cmd.Parameters.Add(parameter);
            cmd.ExecuteNonQuery();
            using (var command = _connection.CreateCommand())
            {
                command.CommandText = "SELECT last_insert_rowid();";
                entity.Id = Convert.ToInt32(command.ExecuteScalar());
            }
            Log.InfoExt("Team saved: " + entity);
            return entity;
        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            _connection.Close();
        }
    }

    public Team? Delete(int id)
    {
        Log.Info($"Deleting team with id {id}");
        Team? team = FindOne(id);
        if (team != null)
        {
            try
            {
                _connection.Open();
                using var cmd = _connection.CreateCommand();
                cmd.CommandText = "DELETE FROM Team WHERE id = @id";
                var parameter = cmd.CreateParameter();
                parameter.ParameterName = "@id";
                parameter.Value = id;
                cmd.Parameters.Add(parameter);
                cmd.ExecuteNonQuery();
                Log.InfoExt("Team deleted: " + id.ToString());
                return team;
            }
            catch (Exception e)
            {
                Log.Error(e);
                throw new Exception(e.Message);
            }
            finally
            {
                _connection.Close();
            }
        }

        return null;
    }

    public Team Update(Team entity)
    {
        Log.Info($"Updating team with id {entity.Id}");
        try
        {
            _connection.Open();
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "UPDATE Team SET Name = @name WHERE id = @id";
            var name = cmd.CreateParameter();
            name.ParameterName = "@name";
            name.Value = entity.Name;
            
            var id = cmd.CreateParameter();
            id.ParameterName = "@id";
            id.Value = entity.Id;
            
            cmd.Parameters.Add(name);
            cmd.Parameters.Add(id);
            cmd.ExecuteNonQuery();
            Log.InfoExt("Team updated: " + entity);
            return entity;
        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            _connection.Close();
        }
    }
}